plugins/wasm-bpf/wasm-bpf.cpp

```diff
@@ -0,0 +1,253 @@
+  return 0;
+}
+
+int bpf_buffer__poll(struct bpf_buffer *buffer, int timeout_ms) {
+  switch (buffer->type) {
+  case BPF_MAP_TYPE_PERF_EVENT_ARRAY:
+    return perf_buffer__poll((perf_buffer *)buffer->inner, timeout_ms);
+  case BPF_MAP_TYPE_RINGBUF:
+    return ring_buffer__poll((ring_buffer *)buffer->inner, timeout_ms);
+  default:
+    return -EINVAL;
+  }
+}
+
+void bpf_buffer__free(struct bpf_buffer *buffer) {
+  if (!buffer)
+    return;
+
+  switch (buffer->type) {
+  case BPF_MAP_TYPE_PERF_EVENT_ARRAY:
+    perf_buffer__free((perf_buffer *)buffer->inner);
+    break;
+  case BPF_MAP_TYPE_RINGBUF:
+    ring_buffer__free((ring_buffer *)buffer->inner);
+    break;
+  }
+  free(buffer);
+}
+/// Get the file descriptor of a map by name.
+int wasm_bpf_program::bpf_map_fd_by_name(const char *name) {
+  return bpf_object__find_map_fd_by_name(obj.get(), name);
+}
+/// @brief load all bpf programs and maps in a object file.
+int wasm_bpf_program::load_bpf_object(const void *obj_buf, size_t obj_buf_sz) {
+  auto object = bpf_object__open_mem(obj_buf, obj_buf_sz, NULL);
+  if (!object) {
+    return (int)libbpf_get_error(object);
+  }
+  obj.reset(object);
+  return bpf_object__load(object);
+}
+
+static int attach_cgroup(struct bpf_program *prog, const char *path) {
+  int fd = open(path, O_RDONLY);
+  if (fd == -1) {
+    printf("Failed to open cgroup\n");
+    return -1;
+  }
+  if (!bpf_program__attach_cgroup(prog, fd)) {
+    printf("Prog %s failed to attach cgroup %s\n", bpf_program__name(prog),
+           path);
+    return -1;
+  }
+  close(fd);
+  return 0;
+}
+
+/// @brief attach a specific bpf program by name and target.
+int wasm_bpf_program::attach_bpf_program(const char *name,
+                                         const char *attach_target) {
+  struct bpf_link *link;
+  if (!attach_target) {
+    link =
+        bpf_program__attach(bpf_object__find_program_by_name(obj.get(), name));
+  } else {
+    struct bpf_object *o = obj.get();
+    struct bpf_program *prog = bpf_object__find_program_by_name(o, name);
+    if (!prog) {
+      printf("get prog %s fail", name);
+      return -1;
+    }
+    const char *sec_name = bpf_program__section_name(prog);
+    // TODO: support more attach type
+    if (strcmp(sec_name, "sockops") == 0) {
+      return attach_cgroup(prog, attach_target);
+    } else {
+      // try auto attach if new attach target is not supported
+      link = bpf_program__attach(
+          bpf_object__find_program_by_name(obj.get(), name));
+    }
+  }
+  if (!link) {
+    return (int)libbpf_get_error(link);
+  }
+  links.emplace(std::unique_ptr<bpf_link, int (*)(bpf_link * obj)>{
+      link, bpf_link__destroy});
+  return 0;
+}
+
+/// polling the buffer, if the buffer is not created, create it.
+int wasm_bpf_program::bpf_buffer_poll(
+    WasmEdge_ExecutorContext *executor,
+    const WasmEdge_ModuleInstanceContext *module_instance, int fd,
+    int32_t sample_func, uint32_t ctx, void *data, size_t max_size,
```

Comment:
If the `sample_func` is used in `uint32_t`, we can change the type to `uint32_t` directly, instead of casting it when using the variable.